tool/python/singa/layer.py

```diff
@@ -0,0 +1,431 @@
+        self.layer.rgbimage_conf.meanfile = meanfile
+
+class Dense(Layer):
+
+    def __init__(self, output_dim=0, activation=None,
+                 init=None, w_param=None, b_param=None, input_dim=None,
+                 **kwargs):
+        '''
+        required
+          output_dim = (int)
+        optional
+          activation = (string)
+          init       = (string)     // 'unirom', 'gaussian', 'constant'
+          w_param    = (Parameter)  // Parameter object for weight
+          b_param    = (Parameter)  // Parameter object for bias
+          **kwargs
+            w_lr = (float) // learning rate for w
+            w_wd = (float) // weight decay for w
+            b_lr = (float) // learning rate for b
+            b_wd = (float) // weight decay for b
+        '''
+        # required
+        assert output_dim > 0, 'output_dim should be set'
+        super(Dense, self).__init__(type=kInnerProduct, **kwargs)
+        self.layer.innerproduct_conf.num_output = output_dim
+        if 'transpose' in kwargs:
+            self.layer.innerproduct_conf.transpose = kwargs['transpose']
+
+        # parameter w (default: gaussian)
+        if w_param == None:
+            self.init = 'gaussian' if init == None else init
+            w_param = Parameter(init=self.init)
+        set_param_field(w_param.param, 'w', False, **kwargs)
+        setval(self.layer, param=w_param.param)
+
+        # parameter b (default: constant)
+        if b_param == None:
+            self.init = 'constant' if init == None else init
+            b_param = Parameter(init=self.init)
+        set_param_field(b_param.param, 'b', False, **kwargs)
+        setval(self.layer, param=b_param.param)
+
+        # following layers: e.g., activation, dropout, etc.
+        if activation:
+            self.mask = Activation(activation=activation).layer
+
+
+''' Classes to deal with multiple layers
+'''
+class Autoencoder(object):
+
+    def __init__(self, hid_dim=None, out_dim=0,
+                 activation=None, param_share=True):
+        '''
+        Generate a set of layers (like MLP) for encoder and decoder
+        The layers are expanded and added in Sequential.add()
+
+        required
+          hid_dim     = (int/list) // the number of nodes in hidden layers
+          out_dim     = (int)      // the number of nodes in the top layer
+        optional 
+          activation  = (string)
+          param_share = (bool)     // to share params in encoder and decoder
+        '''
+
+        # required
+        assert out_dim > 0, 'out_dim should be set'
+        self.out_dim = out_dim
+        assert hid_dim != None, 'hid_dim should be set'
+        self.hid_dim = [hid_dim] if type(hid_dim) == int else hid_dim
+
+        self.layer_type = 'AutoEncoder'
+        self.activation = activation
+        self.param_share = param_share
+
+class RBM(Layer):
+
+    def __init__(self, out_dim=None, w_param=None, b_param=None,
+                 sampling=None, **kwargs):
```

Comment:
similar as Dense and Convolution2D, the w_std, b_lr .. arguments are not included in the argument list nor explained in the comments.